src/main/java/SnippetManager.java

```diff
@@ -86,20 +120,78 @@ void edit(Integer id) {
                                 entry.getValue().getContent().contains(wordToContain)
-                )
+                ) // Match words
                 .filter(entry -> language.equals("") ||
-                        entry.getValue().getLanguage().equals(language))
+                        entry.getValue().getLanguage().equals(language)
+                ) // Match language
                 .filter(entry -> (tags.length == 1 && tags[0].equals("")) ||
-                        Arrays.asList(entry.getValue().getTags()).containsAll(Arrays.asList(tags)))
+                        Arrays.asList(entry.getValue().getTags()).containsAll(Arrays.asList(tags))
+                ) // Match tags
                 .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
     }
 
+    List<Map.Entry<Integer,Snippet>> searchSnippets(String searchTerm) {
+        ArrayList<String> terms = new ArrayList<>(Arrays.asList(searchTerm.split(" ")));
+        terms.forEach(String::toLowerCase);
+        terms.stream().map(term -> term.replaceAll("[^a-zA-Z0-9]", "")).collect(Collectors.toList());
+
+        HashMap<Integer, Integer> matchesPerSnippet = new HashMap<>();
+
+        for (String term : terms) {
+            snippets.entrySet().forEach(entry -> {
+                        Integer key = entry.getKey();
+                        String snippetText = entry.getValue().getContent() + " "
+                                + entry.getValue().getTitle() + " "
+                                + entry.getValue().getLanguage();
+
+                        Integer termOccCount = Math.toIntExact(
+                                Arrays.stream(snippetText.split("[ ,\\.]"))
+                                        .filter(word -> word.equals(term))
+                                        .count());
+
+                        if (matchesPerSnippet.containsKey(key)) {
+                            matchesPerSnippet.put(key, matchesPerSnippet.get(key) + termOccCount);
+                        } else {
+                            matchesPerSnippet.put(key, termOccCount);
+                        }
+                    }
+            );
+        }
+
+        List<Map.Entry<Integer, Integer>> idsByRelevance = new LinkedList<>(matchesPerSnippet.entrySet());
+        idsByRelevance.sort(Map.Entry.comparingByValue());
+        List<Map.Entry<Integer, Snippet>> snippetsByRelevance = new ArrayList<>();
+        for (int i = idsByRelevance.size() - 1; i >= 0; i--) {
+            if (matchesPerSnippet.get(idsByRelevance.get(i).getKey()) > 0) {
+                snippetsByRelevance.add(
+                        new AbstractMap.SimpleEntry<Integer, Snippet>(
+                                idsByRelevance.get(i).getKey(),
+                                snippets.get(idsByRelevance.get(i).getKey())
+                        )
+                );
+            }
+        }
+        return snippetsByRelevance;
+    }
```

Comment:
```suggestion
    List<Map.Entry<Integer,Snippet>> searchSnippets(String searchTerm) {
        ArrayList<String> terms = new ArrayList<>(Arrays.asList(searchTerm.split(" ")));
        terms.forEach(String::toLowerCase);
        terms.stream().map(term -> term.replaceAll("[^a-zA-Z0-9]", "")).collect(Collectors.toList());

        HashMap<Integer, Integer> matchesPerSnippet = new HashMap<>();

        for (String term : terms) {
            snippets.entrySet().forEach(entry -> {
                        Integer key = entry.getKey();
                        String snippetText = entry.getValue().getContent() + " "
                                + entry.getValue().getTitle() + " "
                                + entry.getValue().getLanguage();

                        Integer termOccCount = Math.toIntExact(
                                Arrays.stream(snippetText.split("[ ,\\.]"))
                                        .filter(word -> word.equals(term))
                                        .count());

                        if (matchesPerSnippet.containsKey(key)) {
                            matchesPerSnippet.put(key, matchesPerSnippet.get(key) + termOccCount);
                        } else {
                            matchesPerSnippet.put(key, termOccCount);
                        }
                    }
            );
        }

        List<Map.Entry<Integer, Integer>> idsByRelevance = new LinkedList<>(matchesPerSnippet.entrySet());
        idsByRelevance.sort(Map.Entry.comparingByValue());
        List<Map.Entry<Integer, Snippet>> snippetsByRelevance = new ArrayList<>();
        for (int i = idsByRelevance.size() - 1; i >= 0; i--) {
            if (matchesPerSnippet.get(idsByRelevance.get(i).getKey()) > 0) {
                snippetsByRelevance.add(
                        new AbstractMap.SimpleEntry<Integer, Snippet>(
                                idsByRelevance.get(i).getKey(),
                                snippets.get(idsByRelevance.get(i).getKey())
                        )
                );
            }
        }
        return snippetsByRelevance;
    }
```

Seems solid although I would recommend delegating subroutines to functions rather than including them all in one function.

Reply:
Good point. This should be separated in two methods, one for counting occurrences and another for returning snippets based on that.